[text-underline-position] Make TextUnderlinePosition use less bits

nt1m opened the pull request:
#### 2644fdabea8a7d1f43206f919db3d18882cd88be
<pre>
[text-underline-position] Make TextUnderlinePosition use less bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=277059">https://bugs.webkit.org/show_bug.cgi?id=277059</a>
<a href="https://rdar.apple.com/132452073">rdar://132452073</a>

Reviewed by Matthieu Dubet.

Use an OptionSet so it fits in 4 bits.

Empty OptionSet represents the auto value.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textUnderlinePositionToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::addToLine):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextUnderlinePosition):
(WebCore::RenderStyle::textUnderlinePosition const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTextUnderlinePosition):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/TextUnderlinePosition.cpp: Removed.
* Source/WebCore/rendering/style/TextUnderlinePosition.h: Removed.
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::computedUnderlineOffset):
(WebCore::isAlignedForUnder):
(WebCore::overlineOffsetForTextBoxPainting):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::valueToEmphasisPosition):
(WebCore::Style::valueToUnderlinePosition):
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition):
(WebCore::Style::BuilderConverter::valueToEmphasisPosition): Deleted.

Canonical link: <a href="https://commits.webkit.org/281358@main">https://commits.webkit.org/281358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd2422fc9c9356735985d4582886c9042565b7f5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59661 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63577 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/10185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61790 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46659 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10339 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/10185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61691 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/36404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/51647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/29239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/33108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/8894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9108 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/9173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65308 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3589 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3600 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/51635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/55871 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/13227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/2980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8907 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/34820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/36989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/InlineTextBoxStyle.cpp:
```diff
@@ -157,7 +157,7 @@ static float computedUnderlineOffset(const UnderlineOffsetArguments& context)
     if (isAlignedForUnder(styleToUse)) {
         ASSERT(context.textUnderlinePositionUnder);
         // FIXME: This needs to be flipped for sideways-lr.
-        if (textUnderlinePosition.side == TextUnderlinePosition::Side::Right) {
```
We have a `OptionSet::contains()` which is more readable than a bitwise `&` imho.

mdubet commented on Source/WebCore/style/InlineTextBoxStyle.cpp:
```diff
@@ -157,7 +157,7 @@ static float computedUnderlineOffset(const UnderlineOffsetArguments& context)
         ASSERT(context.textUnderlinePositionUnder);
         // FIXME: This needs to be flipped for sideways-lr.
-        if (textUnderlinePosition.side == TextUnderlinePosition::Side::Right) {
+        if (textUnderlinePosition.contains(TextUnderlinePosition::Right)) {
```
Maybe we should assert here and there that we don't have an invalid state.

`ASSERT(!textUnderlinePosition.contains(TextUnderlinePosition::Left)` here

nt1m replied:
The asserts in the computed style extractor are probably sufficient.

mdubet replied:
The computed style extractor is not used for actual rendering so it's more solid to have the assert here than there

nt1m replied:
Tests would catch it, but I don't feel particularly strongly, I just don't see the benefit of sprinkling assertions around.

nt1m replied:
I'll add it

mdubet commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -879,32 +878,41 @@ inline RefPtr<QuotesData> BuilderConverter::convertQuotes(BuilderState&, const C
-    return {
-        fromCSSValueID<TextUnderlinePosition::Metric>(downcast<CSSPrimitiveValue>(pair->first()).valueID()),
-        fromCSSValueID<TextUnderlinePosition::Side>(downcast<CSSPrimitiveValue>(pair->second()).valueID()),
-    };
```
Not sure it's better, but we can construct it directly.

`return { valueToUnderlinePosition(...pair->first) , valueToUnde...(pair->second) }`

nt1m replied:
That doesn't work because `valueToUnderlinePosition` returns an `OptionSet`, so what you're suggesting would be initializing nested OptionSets

mdubet replied:
Oh yes we need ` return valueToUnder(first) & valueToUnder(second);`, which is definitely not better 😁

nt1m committed:
[text-underline-position] Make TextUnderlinePosition use less bits
https://bugs.webkit.org/show_bug.cgi?id=277059
rdar://132452073

Reviewed by Matthieu Dubet.

Use an OptionSet so it fits in 4 bits.

Empty OptionSet represents the auto value.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textUnderlinePositionToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::addToLine):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextUnderlinePosition):
(WebCore::RenderStyle::textUnderlinePosition const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setTextUnderlinePosition):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/TextUnderlinePosition.cpp: Removed.
* Source/WebCore/rendering/style/TextUnderlinePosition.h: Removed.
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::computedUnderlineOffset):
(WebCore::isAlignedForUnder):
(WebCore::overlineOffsetForTextBoxPainting):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::valueToEmphasisPosition):
(WebCore::Style::valueToUnderlinePosition):
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition):
(WebCore::Style::BuilderConverter::valueToEmphasisPosition): Deleted.

Canonical link: https://commits.webkit.org/281358@main
